app/controllers/stories_controller.rb

```diff
@@ -15,6 +15,7 @@ class StoriesController < ApplicationController
 
   before_action :authenticate_user!, except: %i[index search show]
   before_action :set_cache_control_headers, only: %i[index search show]
+  # before_action :set_json_ld, only: :show
```

Comment:
Unsure of how everyone feels about extracting the `users/show.html.erb` JSON-LD into the `stories_controller.rb` still - thoughts?

Reply:
It seems to me like it should still live in `users/show because, from what I could tell, the stories controller redirects to `users/show`.

Reply:
That's my thought as well @vaidehijoshi !

Reply:
I think it's too hard to build a correct JSON with:

```erb
      <% if @user.twitter_username? %>
        "https://twitter.com/<%= @user.twitter_username %>",
      <% end %>
```

as JSON doesn't tolerate trailing commas so you'll always end up with an invalid document if a key is missing or not. 

This is how I would do it:

```ruby
user_json_ld = {
 "mandatory_key_1": "mandatory_value_1",
 "mandatory_key_2": "mandatory_value_2",
  "sameAs": []
}

user_json_ld["sameAs"].append(@user.mastodon_url) if @user.mastodon_url.present?

# and so on...

# once you've built the final Ruby hash we switch to the view
```

```erb
<script type="application/ld+json">
<%= @user_json_ld.to_json.html_safe %>
</script>
```

You'll see that the document this way passes validation :-)

I tried with:

```ruby
    @user_json_ld = {
      "@context": "http://schema.org",
      "@type": "Person",
      "description": "asdasdsa",
      "disambiguatingDescription": [],
      "image": "/uploads/user/profile_image/11/b8bce798-1962-4477-a203-1ea6c35e72b5.png",
      "mainEntityOfPage": {
        "@id": "https://dev.to/rhymes",
        "@type": "WebPage"
      },
      "name": "rhymes",
      "sameAs": [
        "https://github.com/rhymes",
      ],
      "url": "http://localhost:3000/rhymes",
      "worksFor": [
        {
          "@type": "Organization"
        },
      ]
    }
```

ps. don't build it in a before action, as `show` redirects to many different types of pages, there's a `handle_user_index`, you can build it there before `render template: "users/show"`

Reply:
Agree here too;`handle_user_index` does seem like a better place.